Userland/Libraries/LibWeb/Animations/TimingFunction.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2023, Ali Mohammad Pur <mpfard@serenityos.org>
+ * Copyright (c) 2023, Matthew Olsson <mattco@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/BinarySearch.h>
+#include <LibWeb/Animations/TimingFunction.h>
+#include <math.h>
+
+namespace Web::Animations {
+
+// https://www.w3.org/TR/css-easing-1/#linear-easing-function
+double LinearTimingFunction::operator()(double input_progress, bool) const
+{
+    return input_progress;
+}
+
+static double cubic_bezier_at(double x1, double x2, double t)
+{
+    auto a = 1.0 - 3.0 * x2 + 3.0 * x1;
+    auto b = 3.0 * x2 - 6.0 * x1;
+    auto c = 3.0 * x1;
+
+    auto t2 = t * t;
+    auto t3 = t2 * t;
+
+    return (a * t3) + (b * t2) + (c * t);
+}
+
+// https://www.w3.org/TR/css-easing-1/#cubic-bezier-algo
+double CubicBezierTimingFunction::operator()(double input_progress, bool) const
+{
+    // For input progress values outside the range [0, 1], the curve is extended infinitely using tangent of the curve
+    // at the closest endpoint as follows:
+
+    // - For input progress values less than zero,
+    if (input_progress < 0.0) {
+        // 1. If the x value of P1 is greater than zero, use a straight line that passes through P1 and P0 as the
+        //    tangent.
+        if (x1 > 0.0) {
+            if (x1 == 0.0)
```

Comment:
How can x1 be both >0.0 and ==0.0?

Reply:
I was wondering why the spec didn't handle division by zero... 😅